### ISSUE TITLE
Bug when redeeming a gift code: on shipping form submit, error 500

### ISSUE THREAD

**Issue by janbaykara** [commonknowledge/leftbookclub#122]
```
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]Internal Server Error: /redeem/setup/
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]Traceback (most recent call last):
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/core/handlers/exception.py", line 55, in inner
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    response = get_response(request)
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    response = wrapped_callback(request, *callback_args, **callback_kwargs)
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/views/generic/base.py", line 84, in view
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    return self.dispatch(request, *args, **kwargs)
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/contrib/auth/mixins.py", line 73, in dispatch
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    return super().dispatch(request, *args, **kwargs)
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/views/generic/base.py", line 119, in dispatch
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    return handler(request, *args, **kwargs)
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/views/generic/edit.py", line 153, in post
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    return self.form_valid(form)
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/app/views.py", line 306, in form_valid
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    self.finish_gift_redemption(self.request.session["gift_giver_subscription"])
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/app/views.py", line 329, in finish_gift_redemption
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    create_gift(gift_giver_subscription, self.request.user)
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/app/utils/stripe.py", line 184, in create_gift
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    subscription = stripe.Subscription.create(**args)
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/stripe/api_resources/abstract/createable_api_resource.py", line 22, in create
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    response, api_key = requestor.request("post", url, params, headers)
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/stripe/api_requestor.py", line 122, in request
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    resp = self.interpret_response(rbody, rcode, rheaders)
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/stripe/api_requestor.py", line 399, in interpret_response
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    self.handle_error_response(rbody, rcode, resp.data, rheaders)
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/stripe/api_requestor.py", line 159, in handle_error_response
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]    raise err
2022-04-29T15:21:24Z app[4fbabdde] lhr [info]stripe.error.InvalidRequestError: Request req_gl1REZE1dhW5J3: This customer has no attached payment source or default payment method.
```

**Comment by janbaykara** [comment]
Just making notes about this problem, no questions.

It seems that when redeeming a gift susbcription, sometimes a small charge is attempted:

- Example 1:
  - Gift giver subscription: https://dashboard.stripe.com/test/subscriptions/sub_1Kuy8bKYdS0VccAE6xtDEGZf
  - Gift recipient attempted charge: https://dashboard.stripe.com/test/payments/pi_3KuyA7KYdS0VccAE05XF3RLu

**Comment by melaniepatrick** [comment]
Just tried buying a gift card and got an error message.
I did this on my melanie@leftbookclub account, which is active again as I signed up as a test earlier.
I redeemed a code against my gift card which worked fine.
But when I pressed the final payment button I got this 
![Screenshot 2022-05-04 at 10 14 22](https://user-images.githubusercontent.com/100857719/166654150-cab6b71a-0a75-4656-a06e-c44cb0e07771.png)

**Comment by janbaykara** [comment]
Noted Melanie, I'll take a look at this!

**Comment by janbaykara** [comment]
Good news! This is a new error, specific to production, because we have not yet set up the email server for the gift code emails.

@chrisdevereux is configuring the email server and associated DNS records over at #36, so this won't last too long. We could temporarily disable the email feature in the afternoon to check that this is definitely the issue.

```
2022-05-04T09:14:13Z app[2135da82] lhr [info]Internal Server Error: /welcome/
2022-05-04T09:14:13Z app[2135da82] lhr [info]Traceback (most recent call last):
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/core/handlers/exception.py", line 55, in inner
2022-05-04T09:14:13Z app[2135da82] lhr [info]    response = get_response(request)
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response
2022-05-04T09:14:13Z app[2135da82] lhr [info]    response = wrapped_callback(request, *callback_args, **callback_kwargs)
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/views/generic/base.py", line 84, in view
2022-05-04T09:14:13Z app[2135da82] lhr [info]    return self.dispatch(request, *args, **kwargs)
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/contrib/auth/mixins.py", line 73, in dispatch
2022-05-04T09:14:13Z app[2135da82] lhr [info]    return super().dispatch(request, *args, **kwargs)
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/views/generic/base.py", line 119, in dispatch
2022-05-04T09:14:13Z app[2135da82] lhr [info]    return handler(request, *args, **kwargs)
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/views/generic/base.py", line 185, in get
2022-05-04T09:14:13Z app[2135da82] lhr [info]    context = self.get_context_data(**kwargs)
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/app/views.py", line 125, in get_context_data
2022-05-04T09:14:13Z app[2135da82] lhr [info]    membership_context = self.finish_gift_purchase(
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/app/views.py", line 180, in finish_gift_purchase
2022-05-04T09:14:13Z app[2135da82] lhr [info]    send_mail(
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/core/mail/__init__.py", line 87, in send_mail
2022-05-04T09:14:13Z app[2135da82] lhr [info]    return mail.send()
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/core/mail/message.py", line 298, in send
2022-05-04T09:14:13Z app[2135da82] lhr [info]    return self.get_connection(fail_silently).send_messages([self])
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/core/mail/backends/smtp.py", line 124, in send_messages
2022-05-04T09:14:13Z app[2135da82] lhr [info]    new_conn_created = self.open()
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/app/.venv/lib/python3.9/site-packages/django/core/mail/backends/smtp.py", line 80, in open
2022-05-04T09:14:13Z app[2135da82] lhr [info]    self.connection = self.connection_class(
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/usr/local/lib/python3.9/smtplib.py", line 255, in __init__
2022-05-04T09:14:13Z app[2135da82] lhr [info]    (code, msg) = self.connect(host, port)
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/usr/local/lib/python3.9/smtplib.py", line 341, in connect
2022-05-04T09:14:13Z app[2135da82] lhr [info]    self.sock = self._get_socket(host, port, self.timeout)
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/usr/local/lib/python3.9/smtplib.py", line 312, in _get_socket
2022-05-04T09:14:13Z app[2135da82] lhr [info]    return socket.create_connection((host, port), timeout,
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/usr/local/lib/python3.9/socket.py", line 844, in create_connection
2022-05-04T09:14:13Z app[2135da82] lhr [info]    raise err
2022-05-04T09:14:13Z app[2135da82] lhr [info]  File "/usr/local/lib/python3.9/socket.py", line 832, in create_connection
2022-05-04T09:14:13Z app[2135da82] lhr [info]    sock.connect(sa)
2022-05-04T09:14:13Z app[2135da82] lhr [info]ConnectionRefusedError: [Errno 111] Connection refused
```

**Comment by janbaykara** [comment]
@melaniepatrick I have deployed a fix for this in the meanwhile (bedd1c0bbc93f48da52787581f16525570f926f9) that will gracefully handle any email issues, whilst we install the email server. So we can continue to test the purchase flow in peace.
